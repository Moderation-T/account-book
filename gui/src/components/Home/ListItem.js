import React from 'react';
import { Row, Col } from 'antd';
import { EditOutlined, DeleteOutlined } from '@ant-design/icons';
import PropTypes from 'prop-types';

function ListItem({ item, onModifyItem, onDeleteItem }) {
  // 判断收入还是支出
  // const type = item.category.type === 'outcome' ? '-' : '+';

  return (
    <Row gutter={16}>
      <Col span={4} className="title">
        {item.title}
      </Col>
      <Col span={4} className="price">
        {/* {type} */}
        {item.price}
      </Col>
      <Col span={4} className="date">
        {item.date}
      </Col>
      <Col span={4} className="edit">
        <EditOutlined onClick={onModifyItem} />
      </Col>
      <Col span={4} className="delete">
        <DeleteOutlined onClick={onDeleteItem} />
      </Col>
    </Row>
  );
}

// 类型检查
ListItem.protoTypes = {
  // 函数类型且是必须的
  item: PropTypes.object.isRequired,
  onDeleteItem: PropTypes.func.isRequired,
  onModifyItem: PropTypes.func.isRequired,
};

/* 
// 可以给组件属性设置默认值
ListItem.defaultProps = {
  // 函数类型且是必须的
  onModifyItem: () => {},
};
 */

export default ListItem;
