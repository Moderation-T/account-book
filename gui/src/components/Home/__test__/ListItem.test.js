import React from 'react';
import { shallow } from 'enzyme';
import { Col } from 'antd';
import { EditOutlined, DeleteOutlined } from '@ant-design/icons';

import ListItem from '../ListItem';

/** 测试用例说明
 * 1. 是否正确显示 name price data item 的值
 * 2. 是否成功触发 onModifyItem 与 onDeleteItem 方法
 */

const item = {
  title: 'buy stuff for kitten',
  price: 100,
  date: '2018-08-15',
  monthCategory: '2018-8',
  id: '_kly1klf4g',
  cid: '1',
  timestamp: 1534291200000,
};

const props = {
  item,
  onModifyItem: jest.fn(),
  onDeleteItem: jest.fn(),
};

let wrapper;

describe('test ListItem component', () => {
  //  每次运行单个测试用例的时候会执行这一步
  beforeEach(() => {
    wrapper = shallow(<ListItem {...props}></ListItem>);
  });

  // 创建快照
  it('should render the component to match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should correct render title&price&date value', () => {
    expect(wrapper.find('Row .title').prop('children')).toEqual(item.title);
    expect(wrapper.find('Row .price').prop('children')).toEqual(item.price);
    expect(wrapper.find('Row .date').prop('children')).toEqual(item.date);
  });

  it('should trigger the correct function callbacks', () => {
    wrapper.find(EditOutlined).simulate('click');
    expect(props.onModifyItem).toHaveBeenCalled();
    wrapper.find(DeleteOutlined).simulate('click');
    expect(props.onDeleteItem).toHaveBeenCalled();
  });
});
