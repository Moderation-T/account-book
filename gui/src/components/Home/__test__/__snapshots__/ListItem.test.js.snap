// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ListItem component should render the component to match snapshot 1`] = `ShallowWrapper {}`;
